### PR TITLE
Deprecate Trainer.root_gpu

### LINKED ISSUE #11994: Deprecate `Trainer.root_gpu`
## Proposed refactor

Deprecate this property: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/cf64f3443474a93d23b5afb0417e4a60298006e6/pytorch_lightning/trainer/trainer.py#L1982-L1984

### Motivation

This property is redundant with the Strategy's `root_device` property, which generalizes this check. 
Following https://github.com/PyTorchLightning/pytorch-lightning/pull/11758, this property is not used anywhere else in the framework. This is only used in tests, which can be updated to use `trainer.strategy.root_device` instead

### Pitch

Deprecate this property, and recommend using `trainer.strategy.root_device` instead
Regarding how to deprecate the property, this issue provides an example: https://github.com/PyTorchLightning/pytorch-lightning/issues/10615

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @awaelchli @rohitgr7 @tchaton @akihironitta @kaushikb11 @ananthsub @ninginthecloud @jjenniferdai

### PR DESCRIPTION
**Pull request by vpavankalyan**
## What does this PR do?

Deprecate Trainer.root_gpu and recommends using trainer.strategy.root_device  and replace the test that uses it with Trainer.strategy.root_device

Fixes #11994 

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Yes :smile:

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -2022,7 +2022,8 @@ def num_processes(self) -> int:
 
     @property
     def root_gpu(self) -> Optional[int]:
-        return self._accelerator_connector.root_gpu
+        rank_zero_deprecation("Method `trainer.root_gpu` is deprecated please use `trainer.strategy.root_device`")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2025]: Since this is in a `@property`, we need to increase the stacklevel

```suggestion
        rank_zero_deprecation("Method `trainer.root_gpu` is deprecated please use `trainer.strategy.root_device`", stacklevel=5)
```

**Review by carmocca — COMMENTED** [review]
Can you add a test for the deprecation message in `tests/deprecated_api/test_remove_1-8.py`?

**Comment by ananthsub** [comment]
Hi @vpavankalyan , any updates?

**Comment by vpavankalyan** [comment]
Hi @ananthsub,
Was genuinely held back with other work, but I am free now and will continue to work on this issue.
Thanks for checking in on me.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2022,7 +2022,10 @@ def num_processes(self) -> int:
 
     @property
     def root_gpu(self) -> Optional[int]:
-        return self._accelerator_connector.root_gpu
+        rank_zero_deprecation(
+            "Method `trainer.root_gpu` is deprecated please use `trainer.strategy.root_device`", stacklevel=5
+        )
+        return self._strategy.root_device
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2028]: This used to return the index, so we need to keep that behaviour

```suggestion
        return self._strategy.root_device.index
```

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2022,7 +2022,10 @@ def num_processes(self) -> int:
 
     @property
     def root_gpu(self) -> Optional[int]:
-        return self._accelerator_connector.root_gpu
+        rank_zero_deprecation(
+            "Method `trainer.root_gpu` is deprecated please use `trainer.strategy.root_device`", stacklevel=5
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2026]:
```suggestion
            "Method `trainer.root_gpu` is deprecated please use `trainer.strategy.root_device.index`", stacklevel=5
```

File: tests/models/data/horovod/train_default_model.py [hunk 4]
```diff
@@ -102,7 +102,7 @@ def training_epoch_end(self, outputs) -> None:
     if on_gpu:
         trainer = Trainer(gpus=1, strategy="horovod", max_epochs=1)
         # Test the root_gpu property
-        assert trainer.root_gpu == hvd.local_rank()
+        assert trainer.strategy.root_device == hvd.local_rank()
```
Comment (carmocca) [inline tests/models/data/horovod/train_default_model.py:105]:
```suggestion
        assert trainer.strategy.root_device.index == hvd.local_rank()
```

File: tests/models/test_gpu.py [hunk 5]
```diff
@@ -130,7 +130,7 @@ def test_trainer_num_gpu_0(mocked_device_count_0, gpus, expected_num_gpus, strat
     ],
 )
 def test_root_gpu_property(mocked_device_count, gpus, expected_root_gpu, strategy):
-    assert Trainer(gpus=gpus, strategy=strategy).root_gpu == expected_root_gpu
+    assert Trainer(gpus=gpus, strategy=strategy).strategy.root_device == expected_root_gpu
```
Comment (carmocca) [inline tests/models/test_gpu.py:133]:
```suggestion
    assert Trainer(gpus=gpus, strategy=strategy).strategy.root_device.index == expected_root_gpu
```

File: tests/models/test_gpu.py [hunk 6]
```diff
@@ -142,7 +142,7 @@ def test_root_gpu_property(mocked_device_count, gpus, expected_root_gpu, strateg
     ],
 )
 def test_root_gpu_property_0_passing(mocked_device_count_0, gpus, expected_root_gpu, strategy):
-    assert Trainer(gpus=gpus, strategy=strategy).root_gpu == expected_root_gpu
+    assert Trainer(gpus=gpus, strategy=strategy).strategy.root_device == expected_root_gpu
```
Comment (carmocca) [inline tests/models/test_gpu.py:145]:
```suggestion
    assert Trainer(gpus=gpus, strategy=strategy).strategy.root_device.index == expected_root_gpu
```

File: tests/models/test_gpu.py [hunk 7]
```diff
@@ -142,7 +142,7 @@ def test_root_gpu_property(mocked_device_count, gpus, expected_root_gpu, strateg
     ],
 )
 def test_root_gpu_property_0_passing(mocked_device_count_0, gpus, expected_root_gpu, strategy):
```
Comment (carmocca) [inline tests/models/test_gpu.py:144]:
```suggestion
def test_root_device_property_0_passing(mocked_device_count_0, gpus, expected_root_gpu, strategy):
```

File: tests/models/test_gpu.py [hunk 8]
```diff
@@ -130,7 +130,7 @@ def test_trainer_num_gpu_0(mocked_device_count_0, gpus, expected_num_gpus, strat
     ],
 )
 def test_root_gpu_property(mocked_device_count, gpus, expected_root_gpu, strategy):
```
Comment (carmocca) [inline tests/models/test_gpu.py:132]:
```suggestion
def test_root_device_property(mocked_device_count, gpus, expected_root_gpu, strategy):
```

**Review by carmocca — COMMENTED** [review]
The deprecation test needs to be added to this file, for example, you can copy this one but for `root_gpu`:

https://github.com/PyTorchLightning/pytorch-lightning/blob/15364c18c81909e213f1910a57982f100f8a356d/tests/deprecated_api/test_remove_1-8.py#L164-L169

File: CHANGELOG.md [hunk 9]
```diff
@@ -354,6 +354,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Deprecated `Trainer.lr_schedulers` in favor of `Trainer.lr_scheduler_configs` which returns a list of dataclasses instead of dictionaries ([#11443](https://github.com/PyTorchLightning/pytorch-lightning/pull/11443))
 
+- Deprecated `Trainer.root_gpu` in favor of `Trainer.strategy.root_device` ([#11994](https://github.com/PyTorchLightning/pytorch-lightning/pull/11994))
```
Comment (ananthsub) [inline CHANGELOG.md:357]:
```suggestion
- Deprecated `Trainer.root_gpu` in favor of `Trainer.strategy.root_device.index` ([#11994](https://github.com/PyTorchLightning/pytorch-lightning/pull/11994))
```

**Comment by vpavankalyan** [comment]
As PR #12262 solves the issue, I am closing this Draft PR. 
Being a beginner learned a lot of things and I would like to thank the community for the opportunity and especially @carmocca and @ananthsub for their support.
On to the next one :rocket: :rocket:

**Comment by DuYicong515** [comment]
Sorry I didn't notice that you've worked on this PR already, my bad!
Thanks for working on it @vpavankalyan. Feel free to let me know if you want to continue on this, and I can close mine if you are on it. 

Look forward to your next contributions!
